actors/miner/src/lib.rs

```diff
@@ -2995,6 +2996,301 @@ impl Actor {
+                    params.to_deadline,
+                    policy.wpost_dispute_window
+                ));
+            }
+
+            if !deadline_available_for_move(
+                policy,
+                params.from_deadline,
+                params.to_deadline,
+                current_deadline.index,
+            ) {
+                return Err(actor_error!(
+                    forbidden,
+                    "cannot move to a deadline which is further from current deadline"
+                ));
+            }
+
+            let from_quant = state.quant_spec_for_deadline(policy, params.from_deadline);
+            let to_quant = state.quant_spec_for_deadline(policy, params.to_deadline);
+
+            let mut from_deadline =
+                deadlines.load_deadline(policy, store, params.from_deadline).map_err(|e| {
+                    e.downcast_default(
+                        ExitCode::USR_ILLEGAL_STATE,
+                        format!("failed to load deadline {}", params.from_deadline),
+                    )
+                })?;
+            let mut to_deadline =
+                deadlines.load_deadline(policy, store, params.to_deadline).map_err(|e| {
+                    e.downcast_default(
+                        ExitCode::USR_ILLEGAL_STATE,
+                        format!("failed to load deadline {}", params.to_deadline),
+                    )
+                })?;
+
+            let partitions = &mut params.partitions;
+            if partitions.is_empty() {
+                let partitions_amt = from_deadline.partitions_amt(rt.store()).map_err(|e| {
+                    e.downcast_default(
+                        ExitCode::USR_ILLEGAL_STATE,
+                        format!("failed to load partitions for deadline {}", params.from_deadline),
+                    )
+                })?;
+
+                for partition_idx in 0..partitions_amt.count() {
+                    partitions.set(partition_idx);
+                }
+            }
+
+            let (live, dead, removed_power) =
+                from_deadline.remove_partitions(store, partitions, from_quant).map_err(|e| {
+                    e.downcast_default(
+                        ExitCode::USR_ILLEGAL_STATE,
+                        format!(
+                            "failed to remove partitions from deadline {}",
+                            params.from_deadline
+                        ),
+                    )
+                })?;
+
+            state.delete_sectors(store, &dead).map_err(|e| {
+                e.downcast_default(ExitCode::USR_ILLEGAL_STATE, "failed to delete dead sectors")
+            })?;
+
+            let sectors = state.load_sector_infos(store, &live).map_err(|e| {
+                e.downcast_default(ExitCode::USR_ILLEGAL_STATE, "failed to load moved sectors")
+            })?;
+            let proven = true;
+            let added_power = to_deadline
+                .add_sectors(
```

Comment:
Interesting. This is not what I expected. I can see why you tried it, especially given that `add_sectors` already exists, but this is much more expensive (because it loads sector infos) than literally moving the partition and keeping it intact. It also introduces complications, such as the `proven` parameter. It might work if you could separate the sectors into proven and unproven, but that will require more work lower down.

I think it will be better to keep working at the partition level, copy the partition state directly into the new deadline, then update the deadline metadata with the partition's memoized info directly.

Edit: oh, I see that remove_partition fails if there are any unproven sectors, though that's not documented. It's unfortunate, since it will make it harder for an SP to move partitions while growing, since they don't currently control the deadline for new sectors. But I still think we should more literally just move the partition.